### PR TITLE
Best practices for local development

### PR DESCRIPTION
**Pull request by kaisalmen**
This is mainly a documentation update, plus some build/script additions to `package.json`.

It describes now in detail what steps are required to setup a development environment where you can develop and debug your own language and addition are able to debugging Langium that is itself under development (changed by yourself).

We need to decide if CONTRIBUTING is the best place for the content or if should go somewhere else.

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -4,4 +4,64 @@ Thank you for your interest in the Langium project! The following is a set of gu
 
 ## Prerequisites
 
-Developing Langium only requires `npm@^7.7.0`. For this you will need `Node.js 10` or greater.
\ No newline at end of file
+Developing Langium only requires `npm@^7.7.0`. For this you will need `Node.js 10` or greater.
```
Comment (spoenemann) [inline CONTRIBUTING.md:7]: We should check the actual minimum NodeJS version (see #192). I noticed that we use more recent APIs in some places, so NodeJS 10 certainly doesn't work, but that's fine because it's old and no longer supported.

Looking at https://nodejs.org/en/about/releases/ I think we should use version 12 as minimum version.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -4,4 +4,64 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+
+## Best practices for local development
```
Comment (spoenemann) [inline CONTRIBUTING.md:9]: We should start the contributing guide with the instructions to develop the packages in our repository, and make that as simple as possible, to avoid creating the impression that we somehow need a complex setup. The explanations you added here are useful, but I'd put them into an "advanced" section to clarify that it's not always necessary.

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -4,4 +4,64 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+
+### npm dependency resolution pitfalls
+
+When you add a dependency to a `package.json` npm resolves this from the configured registry (usually https://registry.npmjs.org/). In other dependency management system you can create something like a dev or snapshot version and install it locally, but this is not possible as such with npm. If you want to achieve a similar behavior you need to link your packages globally (global refers to you user's account scope, local refers to the project scope). When you do this you forcefully overwrite what was downloaded from the npm registry. Incrementing the version of Langium itself will led to problems, because npm will not be able to resolve those from its registry and therefore fail any install attempt in projects referring to Langium.
```
Comment (spoenemann) [inline CONTRIBUTING.md:15]:
```suggestion
When you add a dependency to a `package.json` npm resolves this from the configured registry (usually https://registry.npmjs.org/). In other dependency management systems you can create something like a dev or snapshot version and install it locally, but this is not possible as such with npm. If you want to achieve a similar behavior you need to link your packages globally (global refers to you user's account scope, local refers to the project scope). When you do this you forcefully overwrite what was downloaded from the npm registry. Incrementing the version of Langium itself will lead to problems, because npm will not be able to resolve those from its registry and therefore fail any install attempt in projects referring to Langium.
```

File: CONTRIBUTING.md [hunk 4]
```diff
@@ -4,4 +4,64 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+
+Langium is an npm workspace project with currently four main packages (`langium`, `langium-cli`, `langium-vscode` and `generator-langium`). If you execute `npm install` on the top-level of the project all dependencies will be gathered locally inside `node_modules` and all four packages or sub-projects will be build. **Warning**: Do not run npm install directly inside the four Langium packages as it will mess up dependency resolution once you link packages globally.
```
Comment (spoenemann) [inline CONTRIBUTING.md:17]:
```suggestion
Langium is an npm workspace project with currently four main packages (`langium`, `langium-cli`, `langium-vscode` and `generator-langium`). If you execute `npm install` on the top-level of the project all dependencies will be gathered locally inside `node_modules` and all four packages or sub-projects will be built. **Warning**: Do not run npm install directly inside the four Langium packages as it will mess up dependency resolution once you link packages globally.
```

File: package.json [hunk 5]
```diff
@@ -8,7 +8,9 @@
     "lint": "npm run lint --workspaces",
     "test": "npm run test --workspace=langium",
     "langium:generate": "npm run langium:generate --workspace=langium --workspace=domainmodel --workspace=arithmetics --workspace=statemachine",
-    "prepare": "npm run clean && npm run build"
+    "prepare": "npm run clean && npm run build",
+    "dev-build": "npm run dev-clean && npm install && npm link ./packages/langium && npm link ./packages/langium-cli && npm link ./packages/langium-vscode && npm link ./packages/generator-langium",
+    "dev-clean": "rimraf ./packages/**/node_modules && npm uninstall -g langium-workspaces langium-cli langium-vscode generator-langium langium && npm unlink langium-workspaces langium-cli langium-vscode generator-langium langium"
```
Comment (spoenemann) [inline package.json:13]: `langium-vscode` will never be used locally as a package. We don't publish it via npm, but `vsce` / `ovsx`.

File: packages/generator-langium/langium-template/package.json [hunk 6]
```diff
@@ -43,7 +43,7 @@
         "langium:generate": "langium generate"
     },
     "dependencies": {
-        "langium": "^0.1.0",
+        "langium": "0.1.0",
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply packages/generator-langium/langium-template/package.json:46]: I tested a newer version first and this was left over and now over-constrained. I will revert.

File: CONTRIBUTING.md (hunk 1 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:7]: Ok, so minimum version is Maintenance LTS. Are there further requirements apart from workspaces that we have on a specific npm version?

File: CONTRIBUTING.md (hunk 2 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:9]: I agree. I was very verbose on a specific aspect that is maybe only relevant for 10% of future users.
The guide should be simple for the normal use-case which is straightforward and it should make this more obvious (re-structure, simple first, advance use case later).

File: package.json (hunk 5 above)
Comment (kaisalmen) [reply package.json:13]: Ok, then it can be removed here.

File: CONTRIBUTING.md (hunk 3 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:15]: I have already adjusted the test locally and it will be part of an upcoming commit.

File: CONTRIBUTING.md (hunk 4 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:17]: I have already adjusted the test locally and it will be part of an upcoming commit.

File: CONTRIBUTING.md (hunk 1 above)
Comment (spoenemann) [reply CONTRIBUTING.md:7]: No, the workspaces feature is what we need from npm, and regarding NodeJS we need to be careful not to use API that is only available in versions > 12.

File: package.json (hunk 5 above)
Comment (spoenemann) [reply package.json:13]: And what is `langium-workspaces`?

Comment (kaisalmen) [reply package.json:13]: If you perform `npm link` in the root it links the whole langium workspace to the global npm folder. This is some sanity clean-up.

**Comment by kaisalmen** [comment]
I polished the dev guide in CONTRIBUTING once more. I think it is in good shape now.
The three examples now contain READMEs which are linked from the dev guide. These READMEs are mostly stubs for now. In the Arithmetics example I added a little section on how to use the CLI. Idea is that we can extend example README with some little explanation and usage instructions for the future.
We now have a Gitpod button!

**Comment by kaisalmen** [comment]
Commits have been squashed and the `package-lock.json` has been removed.

File: README.md [hunk 7]
```diff
@@ -19,4 +20,5 @@ The main goal of Langium is to lower the barrier of creating a DSL / low-code pl
 
 ## How?
 
-Langium is available as [npm package](https://www.npmjs.com/package/langium). Please read the package documentation to get started.
+Langium is available as [npm package](https://www.npmjs.com/package/langium).
+Please read the [Local Development Guide](./CONTRIBUTING.md#local-development) to get started.
```
Comment (msujew) [inline README.md:24]:
```suggestion
Please read the [Local Development Guide](./CONTRIBUTING.md#local-development-guide) to get started.
```

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+We outline three use-cases:
+1. Developing Langium itself
+2. Developing you own language and use Langium as tool
+3. Developing Langium itself and developing your own language itself.
+
+The first two use-cases are the common ones. The third one is less common, but there are some recommendations you should know.
+
+### Developing Langium
+
+Development of Langium does not necessarily require VSCode (https://code.visualstudio.com/download), but without it you will not be able to use language server and its extension for VSCode.
```
Comment (spoenemann) [inline CONTRIBUTING.md:21]:
```suggestion
Development of Langium does not necessarily require VSCode (https://code.visualstudio.com/download), but without it you will not be able to use the language server and its extension for VSCode.
```

File: CONTRIBUTING.md [hunk 9]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+Langium is an npm workspace project with currently four main packages (`langium`, `langium-cli`, `langium-vscode` and `generator-langium`).
```
Comment (spoenemann) [inline CONTRIBUTING.md:22]: Writing details such as the number and the names of contained packages here means we have to maintain this whenever there are changes. I'd rather shorten this to "multiple packages", and maybe link that to the `packages` folder in GitHub.

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+			"sourceMaps": true,
+			"outFiles": [
+				"${workspaceFolder}/out/**/*.js"
+			],
+			"type": "pwa-node"
+		}
+```
+With the above configuration you can attach the debugger to the language server which effectively means you are able to debug your local version of Langium.
+
+**Hint:** If you want to halt the language server start until you attach the debugger you have to `--inspect=6009` to `--inspect-brk=6009` in the language project's `src/extension.ts` like shown below:
```
Comment (spoenemann) [inline CONTRIBUTING.md:97]:
```suggestion
**Hint:** If you want to halt the language server start until you attach the debugger you have to change `--inspect=6009` to `--inspect-brk=6009` in the language project's `src/extension.ts` like shown below:
```

Comment (spoenemann) [reply CONTRIBUTING.md:97]: We could control this via an environment variable so a code change is not necessary. Could you create an issue for that?

File: examples/arithmetics/README.md [hunk 11]
```diff
@@ -0,0 +1,13 @@
+# Arithmetics Example
+
+## Interpreter CLI
+
+The Arithmetics Example features an interpreter that you can run via cli.
+
+Ensure the complete projects was properly built, otherwise run `npm install` from the root of the Langium project.
+Use `node ./bin/cli.js` to run the cli. Follow the instructions or use `node ./bin/cli.js eval <full-path-to-calc-file>`.
```
Comment (spoenemann) [inline examples/arithmetics/README.md:8]: We should rename `cli.js` to `cli` for consistency.

File: examples/arithmetics/README.md [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+
+## VSCode Extension
+
+Please see use the VSCode run configuration "Run Arithmetics Extension" to launch a new VSCode instance including the extension for this language.
```
Comment (spoenemann) [inline examples/arithmetics/README.md:12]:
```suggestion
Please use the VSCode run configuration "Run Arithmetics Extension" to launch a new VSCode instance including the extension for this language.
```

File: examples/domainmodel/README.md [hunk 13]
```diff
@@ -0,0 +1,4 @@
+# Domain Model Example
+
+Please see use the VSCode run configuration "Run Domainmodel Extension" to launch a new VSCode instance including the extension for this language.
```
Comment (spoenemann) [inline examples/domainmodel/README.md:3]:
```suggestion
Please use the VSCode run configuration "Run Domainmodel Extension" to launch a new VSCode instance including the extension for this language.
```

File: CONTRIBUTING.md [hunk 14]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
\ No newline at end of file
+For developing Langium you require at least the Maintenance LTS version of Node.js (currently 12) and at least npm version 7.7.0 (`npm@^7.7.0`) to be able to use npm workspaces. With Node.js versions below 16 you have to install the correct version of npm with `npm install -g npm@7`.
+
+
+## Local Development Guide
+
+We outline three use-cases:
+1. Developing Langium itself
+2. Developing you own language and use Langium as tool
+3. Developing Langium itself and developing your own language itself.
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply CONTRIBUTING.md:15]: Yes, sorry it should be "in parallel". I will change with next commit.

File: CONTRIBUTING.md [hunk 15]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
-Developing Langium only requires `npm@^7.7.0`. For this you will need `Node.js 10` or greater.
\ No newline at end of file
+For developing Langium you require at least the Maintenance LTS version of Node.js (currently 12) and at least npm version 7.7.0 (`npm@^7.7.0`) to be able to use npm workspaces. With Node.js versions below 16 you have to install the correct version of npm with `npm install -g npm@7`.
+
+
+## Local Development Guide
+
+We outline three use-cases:
+1. Developing Langium itself
+2. Developing you own language and use Langium as tool
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply CONTRIBUTING.md:14]: I will change with next commit.

File: README.md (hunk 7 above)
Comment (kaisalmen) [reply README.md:24]: I will change with next commit.

File: CONTRIBUTING.md (hunk 8 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:21]: I will change it with next commit.

File: CONTRIBUTING.md (hunk 9 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:22]: Good idea. I will change it accordingly.

File: CONTRIBUTING.md [hunk 16]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+
+Take a look at the examples readily available inside the repository to get started:
+* [Arithmetics](./examples/arithmetics/README.md)
+* [Domain Model](./examples/domainmodel/README.md)
+* [State Machine](./examples/statemachine/README.md)
+
+
+### Developing your own language
+
+The Please follow the instructions outlined in the [Langium package documentation](./packages/langium/README.md) to get started.
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply CONTRIBUTING.md:33]: I will change it accordingly

File: CONTRIBUTING.md [hunk 17]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+
+
+### Developing Langium and your own language in parallel
+
+If for whatever reason you want:
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply CONTRIBUTING.md:38]: I will change it accordingly

File: CONTRIBUTING.md [hunk 18]
```diff
@@ -4,4 +4,97 @@ Thank you for your interest in the Langium project! The following is a set of gu
...
+- Develop or change Langium locally
+- Use your Langium changes in your own language project (independent of Langium repository)
+- Debug all code in all projects
+then you should follow some recommendations.
+
+#### npm dependency resolution pitfalls
+
+If you are have long experience in using node and npm this sub-chapter will likely offer no new info an you can skip it. But, if you are more familiar with other languages, build and dependency management eco-systems then this sub-chapter could be helpful to you.
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply CONTRIBUTING.md:46]: I will change it accordingly

File: CONTRIBUTING.md (hunk 10 above)
Comment (kaisalmen) [reply CONTRIBUTING.md:97]: I will change it accordingly and I just created #215

File: examples/arithmetics/README.md (hunk 11 above)
Comment (kaisalmen) [reply examples/arithmetics/README.md:8]: Do you want to open a new issue for that?

File: examples/arithmetics/README.md [hunk 19]
```diff
@@ -0,0 +1,13 @@
+# Arithmetics Example
+
+## Interpreter CLI
+
+The Arithmetics Example features an interpreter that you can run via cli.
+
+Ensure the complete projects was properly built, otherwise run `npm install` from the root of the Langium project.
```
[the remark this answers is not in the document]
Comment (kaisalmen) [reply examples/arithmetics/README.md:7]: I will change it accordingly

File: examples/arithmetics/README.md (hunk 12 above)
Comment (kaisalmen) [reply examples/arithmetics/README.md:12]: Yes, I createdI will change it accordingly

File: examples/domainmodel/README.md (hunk 13 above)
Comment (kaisalmen) [reply examples/domainmodel/README.md:3]: I will change it accordingly

**Comment by kaisalmen** [comment]
I covered all comments and pushed an update.

**Comment by kaisalmen** [comment]
Force push was just about `--save` / `--save-dev` is no longer needed when using `npm install`

File: CONTRIBUTING.md [hunk 20]
```diff
@@ -56,8 +55,8 @@ There is an npm build target available (`npm run dev-build`) linking all Langium
 A project you created with `yo langium` contains a dependency to `langium` (e.g. `0.1.0`) and a dev-dependency to `langium-cli`) inside `package.json` by default. Now, the easiest way to switch to your own build is to locally reference the required Langium packages folders.
 Let's assume your local Langium checkout resides on the same level as your own language project, then these two commands will set the references inside the `package.json`:
 ```shell
-npm install --save ../langium/packages/langium
-npm install --save-dev ../langium/packages/langium-cli
+npm install ../langium/packages/langium
+npm install -D ../langium/packages/langium-cli
```
Comment (sailingKieler) [inline CONTRIBUTING.md:59]: I figured out that can use `npm link -S langium` and `npm link -D langium-cli` to add those dependencies in the versions "installed"/"published" to the user global package cache. If a package is installed as link to a local folder via `npm link` the link target is then referenced in the `package.json`. Otherwise, the path of the package within the global package cache is used. If the particular package is not available in the global cache it will be downloaded from the public registry first.
In contrast to the statement in the `help` contents (seem to be equal for several commands) saying `--safe` is true by default, this is not the case for me. I had to state `--safe` or `-S` get the dependency added in the `package.json`.

I suggest to replace the above statements by the `npm link`-based ones as they appear to be more consistent with the preparations suggested above to me.

Comment (dhuebner) [reply CONTRIBUTING.md:59]: What does -S and -D mean?

Comment (sailingKieler) [reply CONTRIBUTING.md:59]: -S: "add the package to the 'dependencies' list in the package.json" (since npm5 by default true for `npm install`)
-D: "add the package to the 'devDependencies' list in the package.json"

Comment (kaisalmen) [reply CONTRIBUTING.md:59]: I am pretty certain that you need the local reference for the VSCode debugger to work correctly in the VSCode workspace, but I may be in error here. We should check this before altering the description.

Comment (kaisalmen) [reply CONTRIBUTING.md:59]:
```shell
npm link -S langium
npm link -D langium-cli
```
creates the local file references as desired (my local build was globally installed before (machine-wide). So, this can replace the other instruction then. @sailingKieler It seems you can only omit the options in case of `npm install` and not in case of `npm link`. I can confirm your observation.

File: examples/arithmetics/README.md (hunk 11 above)
Comment (msujew) [reply examples/arithmetics/README.md:8]: You are allowed to omit the `.js` from files, when invoking them using `node`. So you can omit this from the documentation here as well. Removing the file extension from the file in the repo itself can be done separately.

Comment (kaisalmen) [reply examples/arithmetics/README.md:8]: Ok, good to know. We should do the renaming once all examples have CLIs, right?

Comment (kaisalmen) [reply examples/arithmetics/README.md:8]: Means: I will only update the documentation for now.

Comment (msujew) [reply examples/arithmetics/README.md:8]: Exactly, yes 👍

Comment (kaisalmen) [reply examples/arithmetics/README.md]: Done, see squashed commit below

File: README.md [hunk 21]
```diff
@@ -19,4 +20,5 @@ The main goal of Langium is to lower the barrier of creating a DSL / low-code pl
 
 ## How?
 
-Langium is available as [npm package](https://www.npmjs.com/package/langium). Please read the package documentation to get started.
+Langium is available as [npm package](https://www.npmjs.com/package/langium).
+Please read the [Local Development Guide](./CONTRIBUTING.md#local-development-guide) to get started.
```
Comment (spoenemann) [inline README.md:24]: I wouldn't direct all users to that guide, only those who want to go deeper with the internals of Langium.
```suggestion

Please read the [Contribution Guide](./CONTRIBUTING.md) if you want to contribute to Langium.
```

**Review by spoenemann — APPROVED** [review]
Last suggestion from me, otherwise it's :+1:

**Comment by kaisalmen** [comment]
@spoenemann I took your last suggestion on-board and squashed everything.
